### PR TITLE
Empty t log

### PR DESCRIPTION
**Pull request by Stuart-Moore**
Fixes # 

Changes proposed in this pull request:
 -  Copes with the "empty" backup issues
 - Fixed the missing diff problem (just runs the full log chain if it exisits)
 - Added more pestering to make sure we don't reintroduce the above
 - Fixed up some of the issues introduced with the new version of Get-DbaBackupHistory. Needs some love on that end to completely fix (sep issue going in) 

Looks like a lot of files, but 4 are the new tests (2 test scripts and 2 object defs)

How to test this code: 
- [ ] 
- [ ] 

Has been tested on minimum requirements:
- [ ]  Powershell 3
- [ ]  Windows 7
- [ ]  SQL Server 2000

Has been tested on maximum requirements:
- [ ]  SQL Server vNext
- [ ]  Windows 10
- [ ]  Azure Database

Tests for tester:
- [ ] Working/useful help content, including link to command on dbatools web site
- [ ] All examples work as advertised
- [ ] Does not contain template content
- [ ] Does not contain excessive/unnecessary amounts of comments
- [ ] Works remotely
- [ ] Works locally
- [ ] Works on lower versions or throws error specifying version not supported
- [ ] Works with named instances
- [ ] Works with clustered instances
- [ ] Handles offline/read only databases
- [ ] Supports multiple servers (at the command line or piped from Get-SqlRegisteredServerName)
- [ ] No un-handled errors which stop the command working with multiple servers

### CODE REVIEW

**Comment by Stuart-Moore** [comment]
Oops, appveyor doesn't work how I thought it did. Let me work it out

**Comment by potatoqualitee** [comment]
Looks great! Will merge when tests are finished.

**Comment by Stuart-Moore** [comment]
Looks like my stuff is passing now and the errors are from some of theother tests:
https://ci.appveyor.com/project/ctrlbold/dbatools/build/0.8.3387/tests

**Comment by Stuart-Moore** [comment]
@SQLDBAWithABeard Hi Rob, can you see anything wrong here? My pester stuff is passing fine, this appears to be an issue with other people's tests.

**Comment by Zhunya** [comment]
@Stuart-Moore Please, can you make changes for better handling $IgnoreLog?
those I made in closed #1014

**Comment by SQLDBAWithABeard** [comment]
You are correct @Stuart-Moore the failures are for other tests. Cant see what is going on with a quick look

**Comment by Stuart-Moore** [comment]
Cool, not me then :) Playing around locally puts most of the blame on Out-DataTable.Tests.ps1, which doesn't appear to follow any of the usual dbatools/appveyore/cookiemonster  standards. The others appears to be ErrorActionPreference stopping tests

@Zhunya - I'll check the log skipping again. This was where you wanted to be able to just restore the full and the diff, then stop?

**Comment by Zhunya** [comment]
@Stuart-Moore this is where we can speed up restore process. So -IgnoreLog switch will work as intended - skipping logs.

**Comment by Stuart-Moore** [comment]
Do you mean skipping the restores or the scan? Skipping the restores works (or does for me here):
image

(2nd output doesn't have the 5 log restores the first does.

**Comment by Zhunya** [comment]
@Stuart-Moore skipping them on the scan, rocket speed restore process

**Comment by Stuart-Moore** [comment]
Catch 22,we don't know they're log files till we've scanned then.......

**Comment by Zhunya** [comment]
@Stuart-Moore ok, when you have some directory with bunch of files - yes, do not skip log scan, but when you have OLA's style hierarchy, you can freely skip log files!
I propose golden mean - when ```-MaintanaceSolution``` specified with ```-IgnoreLog``` you don't add files in LOG folder to ```$Results``` in ```Get-OlaHRestoreFile.ps1```

**Comment by Stuart-Moore** [comment]
That would work.

I've been to places with all sorts of file naming conventions (one didn't have any file extensions for some reason) so I'm wary of making assumptions. 

That'll go in the feature release coming next. This was just a bug fix, but got held up with the pester stuff. 

@potatoqualitee mr sewell says its someone else's pester broken, so should be OK to merge.

**Comment by Zhunya** [comment]
@Stuart-Moore PLEASE PUT closing ```}``` on line 316 in file Restore-DbaDatabase.ps1!!!

File: functions/Restore-DbaDatabase.ps1 [hunk 1]
```diff
@@ -305,10 +305,16 @@ folder for those file types as defined on the target instance.
...
 				{
-					$f = $f | Select-Object *,  @{Name="Type";Expression={"Full"}}
+					#$f = $f | Select-Object *,  @{Name="Type";Expression={"Full"}}
 				}
+				if ("BackupSetGUID" -notin $f.PSobject.Properties.name)
+				{
+					#This line until Get-DbaBackupHistory gets fixed
+					$f = $f | Select-Object *, @{Name="BackupSetGUID";Expression={$_.BackupSetupID}}
+					#This one once it's sorted:
+					#$f = $f | Select-Object *, @{Name="BackupSetGUID";Expression={$_.BackupSetID}}
```
Comment (Zhunya) [inline functions/Restore-DbaDatabase.ps1]: Please, put } on 316 line!

Comment (Zhunya) [inline functions/Restore-DbaDatabase.ps1]: Oh, sorry for double!

File: functions/Restore-DbaDatabase.ps1 [hunk 2]
```diff
@@ -305,10 +305,16 @@ folder for those file types as defined on the target instance.
...
 
-				$BackupFiles += $F | Select-Object *, @{Name="ServerName";Expression={$_.SqlInstance}}, @{Name="BackupStartDate";Expression={$_.Start}}
+				$BackupFiles += $F | Select-Object *, @{Name="ServerName";Expression={$_.SqlInstance}}, @{Name="BackupStartDate";Expression={$_.Start -as [DateTime]}}
 				$str = ($BackUpFiles | select Fullname) -join ','
```
Comment (Zhunya) [inline functions/Restore-DbaDatabase.ps1]: '''Select-Object''' instead of just '''select'''?

**Comment by Stuart-Moore** [comment]
Sorry, forgot say last commit on thursday fixed @Zhunya issue

So all good to go

**Comment by potatoqualitee** [comment]
merged yall 👍

**Pull request merged**